tsfresh/feature_extraction/data.py

```diff
@@ -0,0 +1,343 @@
+import itertools
+from collections import namedtuple
+from typing import Generator, Iterable, Sized
+
+import pandas as pd
+
+
+class Timeseries(namedtuple('Timeseries', ['id', 'kind', 'data'])):
+    """
+    Timeseries tuple used for feature extraction
+    """
+
+
+class TsData(Iterable[Timeseries], Sized):
+    """
+    TsData provides access to time series data for internal usage.
+
+    Implementations must at least overwrite `__iter__` which must yield tuples containing
+    (id, kind, pd.Series). Make sure `kind` is of type `str` to allow inference
+    of feature settings in `feature_extraction.settings.from_columns`.
+
+    Other methods should be overwritten if a more efficient solution exists for the underlying data store.
+    """
+
+    def __iter__(self):
+        raise NotImplementedError
+
+    def __len__(self):
+        return sum(1 for _ in self)
```

Comment:
For your example: I think this would be a wrong implementation of TsData. An iterator is an iterable, sure, but it would not mach the contract of the class.
But I also thought about this method for a different reason: Even if it's possible to have a default which "works", this is heavily inefficient and will probably never be used (and so far it isn't). I will make it abstract (raise an error).